### PR TITLE
Added new Managing Huge Pages topic

### PR DESCRIPTION
**Pull request by ahardin-rh**
Preview Build: http://file.rdu.redhat.com/~ahardin/01052018/hugepages/scaling_performance/managing_hugepages.html  

https://trello.com/c/steXZKC4/638-5-psap-hugepages-documentation

https://trello.com/c/zGc0NTZE/756-scalehugepages-documentation

### CODE REVIEW

**Comment by ahardin-rh** [comment]
@jeremyeder PTAL

File: scaling_performance/managing_hugepages.adoc [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+[[huge-pages-prerequisites]]
+== Prerequisites
+
+. Nodes must pre-allocate huge pages in order for the node to report its huge
+ page capacity. A node can only pre-allocate huge pages for a single size.
+
+. `HugePages` must be set to `true` across the system:
++
+----
+--feature-gates=HugePages=true
```
Comment (jeremyeder) [inline scaling_performance/managing_hugepages.adoc]: This is the kubelet argument format. but in openshift we have the yaml-based kubelet config. please refer back to the cpu manager docs for an example of how to set a feature gate in the node-config.yaml.  Then replace this line with the yaml example.

File: scaling_performance/managing_hugepages.adoc [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+`hugepages-2Mi`. Unlike CPU or memory, huge pages do not support overcommitment.
+
+----
+apiVersion: v1
+kind: Pod
+metadata:
+  generateName: hugepages-volume-
+spec:
+  containers:
+  - image: fedora:latest
```
Comment (jeremyeder) [inline scaling_performance/managing_hugepages.adoc]: probably make this rhel7

File: scaling_performance/managing_hugepages.adoc [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    command:
+    - sleep
```
Comment (jeremyeder) [inline scaling_performance/managing_hugepages.adoc]: the cpu manager examples use the pause pod.  we should probably unify these examples.  i suppose sleep infinity is the easiest since it is in the rhel7 base image.

File: scaling_performance/managing_hugepages.adoc [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    - inf
+    name: example
+    volumeMounts:
+    - mountPath: /hugepages
+      name: hugepage
+    resources:
+      limits:
+        hugepages-2Mi: 100Mi
```
Comment (jeremyeder) [inline scaling_performance/managing_hugepages.adoc]: I think we need a call-out box related to this line.  Users of hugepages are used to specifying the amount of memory for hugepages a quantity of hugepages, multiplied by the size of the page.  They do this intuitively and it's been around for decades.

What we are doing here is different. We are allowing the user to specify the amount of memory for hugepages as the exact amount to be allocated.  So there is no "multiplied by the size of the page" step mental calculation.

In other words, given a hugepage size of 2MB, if a user wanted to use 100MB of hugepage-backed RAM for their application, they would allocate 50 hugepages.  Above you see we handle the math for users, allowing them to specify 100MB directly.

File: scaling_performance/managing_hugepages.adoc [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+----
+
+* Huge page requests must equal the limits. This is the default if limits are
+specified, but requests are not.
+* Huge pages are isolated at a pod scope, container isolation is planned in a
+future iteration.
+* `EmptyDir` volumes backed by huge pages must not consume more huge page memory
+than the pod request.
+* Applications that consume huge pages via `shmget()`` with `SHM_HUGETLB` must run
+with a supplemental group that matches *_proc/sys/vm/hugetlb_shm_group_*.
```
Comment (jeremyeder) [inline scaling_performance/managing_hugepages.adoc]: We probably need an example of this.

**Review by jeremyeder — CHANGES_REQUESTED** [review]

**Comment by jeremyeder** [comment]
I also want to add some context as to why this exists (from the original hugepages design doc for kube).  I will take a stab at that next week.

**Comment by jeremyeder** [comment]
@derekwaynecarr fyi, have you an example of the supplemental group part?

**Comment by ahardin-rh** [comment]
@jeremyeder I made some updates and yoinked some of Derek's content on the purpose. 

Updated Preview Build (You may need to clear cache to see latest changes): http://file.rdu.redhat.com/~ahardin/01052018/hugepages/scaling_performance/managing_hugepages.html

**Comment by ahardin-rh** [comment]
@derekwaynecarr ^^

File: scaling_performance/managing_hugepages.adoc [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+node supports 2048KiB page sizes, it will expose a schedulable resource
+`hugepages-2Mi`. Unlike CPU or memory, huge pages do not support overcommitment.
+
+----
+apiVersion: v1
+kind: Pod
+metadata:
+  generateName: hugepages-volume-
+spec:
+  containers:
```
Comment (derekwaynecarr) [inline scaling_performance/managing_hugepages.adoc]: add the following:

```
containers:
 securityContext:
  privileged: true
```
I suspect on the majority of machines in the world will report `cat /proc/sys/vm/hugetlb_shm_group` as `0`, so you effectively need to run privileged right now.

**Comment by ahardin-rh** [comment]
@derekwaynecarr Updated with the latest. Thanks for your help!

**Comment by ahardin-rh** [comment]
@jeremyeder @derekwaynecarr Is this good to go for 3.9? Are any further changes needed to the docs? Thanks!

**Comment by jeremyeder** [comment]
lgtm

**Comment by ahardin-rh** [comment]
@wjiangjay Can you please give a QE review? Thanks!

File: scaling_performance/managing_hugepages.adoc [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    securityContext:
+      privileged: true
+  - image: rhel7:latest
```
Comment (wjiangjay) [inline scaling_performance/managing_hugepages.adoc]: here should be 
```
kind: Pod
metadata:
  generateName: hugepages-volume-
spec:
  containers:
  - securityContext:
      privileged: true
    image: rhel7:latest
    command:
    - sleep
    - inf
    name: example
    volumeMounts:
    - mountPath: /hugepages
      name: hugepage
    resources:
      limits:
        hugepages-2Mi: 100Mi 
  volumes:
  - name: hugepage
    emptyDir:
      medium: HugePages
```

File: scaling_performance/managing_hugepages.adoc [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
++
+----
+# cat /etc/origin/node/node-config.yaml
+...
+kubeletArguments:
+...
+  feature-gates:
+  - HugePages=true
+
+# systemctl restart atomic-openshift-node
```
Comment (wjiangjay) [inline scaling_performance/managing_hugepages.adoc]: Here also need master-api enabled for feature-gates "HugePages=true"

**Comment by wjiangjay** [comment]
Others LGTM, and do we need the section "how to switch hugepage size"?

**Comment by ahardin-rh** [comment]
@wjiangjay Thanks! Updated.

@jeremyeder @derekwaynecarr Did you want to add information on switching hugepage size per the comment from QE? If so, let me know what details are missing. Thanks!

**Comment by jeremyeder** [comment]
Meaning an additional fragment that demonstrates

hugepages-1Gi: 2Gi

And perhaps a link out to the RHEL docs on it:
https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/performance_tuning_guide/s-memory-transhuge

**Comment by ahardin-rh** [comment]
^^ @wjiangjay Can you clarify what is needed in regard to switching hugepage size?

**Comment by wjiangjay** [comment]
@ahardin-rh  
```
Some platforms support multiple huge page sizes.  To allocate huge pages
of a specific size, one must precede the huge pages boot command parameters
with a huge page size selection parameter "hugepagesz=<size>".  <size> must
be specified in bytes with optional scale suffix [kKmMgG].  The default huge
page size may be selected with the "default_hugepagesz=<size>" boot parameter.
```
refer to https://www.kernel.org/doc/Documentation/vm/hugetlbpage.txt

**Comment by wjiangjay** [comment]
@ahardin-rh  comment from @jeremyeder is suitable, just need make a reference to the RHEL docs if we want this.

File: scaling_performance/managing_hugepages.adoc (hunk 8 above)
Comment (wjiangjay) [reply scaling_performance/managing_hugepages.adoc]: @ahardin-rh  Do not forget this.

**Comment by ahardin-rh** [comment]
Note to self :spiral_notepad: 
I need to add this: https://github.com/redhat-performance/openshift-psap/blob/master/playbooks/roles/cpumanager-hugepages/tasks/main.yaml#L28

**Comment by ahardin-rh** [comment]
@jeremyeder I made some updates PTAL

**Comment by jeremyeder** [comment]
The update around master-config looks good.  I see now that @wjiangjay spotted it 16 days ago :-)

File: scaling_performance/managing_hugepages.adoc [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+----
+
+. The master API must be enabled for `feature-gates` with `HugePages=true`:
++
+----
+# cat /etc/origin/master/master-config.yaml
+...
+kubeletArguments:
```
Comment (wjiangjay) [inline scaling_performance/managing_hugepages.adoc]: Sorry about that, seems I make you misunderstand, 

here shoud be 
``
kubernetesMasterConfig:
  apiServerArguments:
  ...
  feature-gates:
  - HugePages=true
``
